contracts/SeriesIndexer.sol

```diff
@@ -0,0 +1,792 @@
+    ) external returns (uint256) {
+        return _createSeries(artistAddrs, metadataURI, tokenIDsMapURI);
+    }
+
+    /**
+     * @dev Batch creation of series
+     */
+    function batchAddSeries(
+        address[][] calldata artistsArray,
+        string[] calldata metadataURIs,
+        string[] calldata tokenIDsMapURIs
+    ) external returns (uint256[] memory) {
+        _validateBatchParameters(artistsArray.length, metadataURIs, tokenIDsMapURIs);
+
+        uint256[] memory createdSeriesIDs = new uint256[](artistsArray.length);
+        for (uint256 i = 0; i < artistsArray.length; i++) {
+            createdSeriesIDs[i] = _createSeries(
+                artistsArray[i],
+                metadataURIs[i],
+                tokenIDsMapURIs[i]
+            );
+        }
+        return createdSeriesIDs;
+    }
+
+    /**
+     * @dev Updates an existing series' metadata and token data
+     */
+    function updateSeries(
+        uint256 seriesID,
+        string calldata metadataURI,
+        string calldata tokenIDsMapURI
+    ) external {
+        _updateSeries(seriesID, metadataURI, tokenIDsMapURI);
+    }
+
+    /**
+     * @dev Updates an existing series' metadata and token data
+     */
+    function batchUpdateSeries(
+        uint256[] calldata seriesIDs,
+        string[] calldata metadataURIs,
+        string[] calldata tokenIDsMapURIs
+    ) external {
+        _validateBatchParameters(seriesIDs.length, metadataURIs, tokenIDsMapURIs);
+        for (uint256 i = 0; i < seriesIDs.length; i++) {
+            _updateSeries(seriesIDs[i], metadataURIs[i], tokenIDsMapURIs[i]);
+        }
+    }
+
+    /**
+     * @dev Deletes series and cleans up related data
+     */
+    function deleteSeries(uint256 seriesID) external {
+        _deleteSeries(seriesID);
+    }
+
+    /**
+     * @dev Batch deletes series and cleans up related data
+     */
+    function batchDeleteSeries(uint256[] calldata seriesIDs) external {
+        for (uint256 i = 0; i < seriesIDs.length; i++) {
+            _deleteSeries(seriesIDs[i]);
+        }
+    }
+
+    // ============ Public/External Functions: Series Artist Management ============
+
+    /**
+     * @dev Allows an artist to remove themselves from a series
+     */
+    function removeSelfFromSeries(uint256 seriesID) external seriesExists(seriesID) onlyArtist(seriesID) {
+        uint256 artistID = addressToArtistID[msg.sender];
+
+        Series storage series = seriesDetails[seriesID];
+        uint256[] storage artistIDs = series.artistIDs;
+        
+        // Find and remove artistID from series' artist list
+        for (uint256 i = 0; i < artistIDs.length; i++) {
+            if (artistIDs[i] == artistID) {
+                artistIDs[i] = artistIDs[artistIDs.length - 1];
+                artistIDs.pop();
+                break;
+            }
+        }
+
+        // Remove series from artist's list
+        _removeSeriesFromArtist(artistID, seriesID);
+
+        emit SeriesUpdated(
```

Comment:
Same [here](https://github.com/bitmark-inc/feralfile-exhibition-smart-contract/pull/44#discussion_r1905029940). If you just want to fire a general event for series update, you might just want to pass the `seriesID` and the client should use the seriesID to query the details from the contract. Adding more unrelated parameters in the event could cost more gas and also make it more confusing.